File: runtime/peerbot/src/com/newbound/p2p/protocol/PUBKEY.java
package com.newbound.p2p.protocol;

import org.json.JSONObject;

import com.newbound.p2p.Codes;
import com.newbound.p2p.P2PPeer;
import com.newbound.p2p.P2PResponse;
import com.newbound.p2p.P2PService;
import com.newbound.p2p.P2PSocket;
import com.newbound.robot.Callback;

public class PUBKEY implements Callback {

	P2PService P2P;

	public PUBKEY(P2PService p2p) 
	{
		super();
		P2P = p2p;
	}

	@Override
	public void execute(JSONObject data) 
	{
		try
		{
			String uuid = data.getString("uuid");
			byte[] ba = (byte[])data.get("data");
			P2PSocket sock = (P2PSocket)data.get("p2psocket");
			P2PPeer p = P2P.getPeer(uuid);
			System.out.println("Got public key "+uuid);
			if (p.getPublicKey() == null) 
			{
				p.setPublicKey(ba);
//				sock.CONNECTING = false;
				
				if (p.getReadKey() == null)
				{
					sock.CONNECTING = true;
					sock.send(new P2PResponse(Codes.SEND_READKEY, "".getBytes()));
				}
				
			}
			else System.out.println("Already had a public key!!! Ignoring.");
			P2P.isOK(uuid, sock, null, null); 
		}
		catch (Exception x) { x.printStackTrace(); }
	}

}

File: runtime/peerbot/src/newbound/p2p/protocol/pubkey.py
import json
from newbound.p2p.p2presponse import P2PResponse
import newbound.p2p.codes as Codes

class PUBKEY(object):
    def __init__(self, service):
        self.service = service

    def execute(self, data):
        uuid = data['uuid']
        ba = data['data']
        sock = data['p2psocket']
        p = self.service.p2p.getPeer(uuid)
        if p.pub != None:
            print("Received pubkey for "+uuid+" but we already have one. Ignoring.")
        else:
            p.pub = ba
            self.service.p2p.peers.savePeer(p)
        if p.readkey == None:
            sock.connecting = True
            #FIXME - should not need to send pubkey
            sock.parser.send(P2PResponse(Codes.PUBKEY, self.service.p2p.pub))
            sock.parser.send(P2PResponse(Codes.SEND_READKEY, b''))
        self.service.p2p.isOK(uuid, sock, None, None)
